x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/RateIntAggregator.java

```diff
@@ -0,0 +1,277 @@
+import org.elasticsearch.common.util.ObjectArray;
+import org.elasticsearch.compute.ann.GroupingAggregator;
+import org.elasticsearch.compute.ann.IntermediateState;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.DoubleBlock;
+import org.elasticsearch.compute.data.DoubleVector;
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.LongBlock;
+import org.elasticsearch.compute.operator.DriverContext;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+
+/**
+ * A rate grouping aggregation definition for int.
+ * This class is generated. Edit `X-RateAggregator.java.st` instead.
+ */
+@GroupingAggregator(
+    includeTimestamps = true,
+    value = {
+        @IntermediateState(name = "timestamp", type = "LONG_BLOCK"),
+        @IntermediateState(name = "value", type = "INT_BLOCK"),
+        @IntermediateState(name = "compensation", type = "DOUBLE") }
+)
+public class RateIntAggregator {
+    public static IntRateGroupingState initGrouping(BigArrays bigArrays, long unitInMillis) {
+        // TODO: pass BlockFactory instead bigArrays so we can use the breaker
+        return new IntRateGroupingState(bigArrays, unitInMillis);
+    }
+
+    public static void combine(IntRateGroupingState current, int groupId, long timestamp, int value) {
+        current.append(groupId, timestamp, value);
+    }
+
+    public static void combineIntermediate(
+        IntRateGroupingState current,
+        int groupId,
+        LongBlock timestamps,
+        IntBlock values,
+        double compensation,
+        int otherPosition
+    ) {
+        current.combine(groupId, timestamps, values, compensation, otherPosition);
+    }
+
+    public static void combineStates(
+        IntRateGroupingState current,
+        int currentGroupId,//
+        IntRateGroupingState state,
+        int statePosition
+    ) {
+        throw new UnsupportedOperationException("ordinals grouping is not supported yet");
+    }
+
+    public static Block evaluateFinal(IntRateGroupingState state, IntVector selected, DriverContext driverContext) {
+        return state.evaluateFinal(selected, driverContext.blockFactory());
+    }
+
+    private static class IntRateState implements Accountable {
+        static final long BASE_RAM_USAGE = RamUsageEstimator.sizeOfObject(IntRateState.class);
+        final long[] timestamps; // descending order
+        final int[] values;
+        double compensation = 0;
+
+        IntRateState(int initialSize) {
+            this.timestamps = new long[initialSize];
+            this.values = new int[initialSize];
+        }
+
+        IntRateState(long[] ts, int[] vs) {
+            this.timestamps = ts;
+            this.values = vs;
+        }
+
+        private int dv(int v0, int v1) {
+            return v0 > v1 ? v1 : v1 - v0;
+        }
```

Comment:
I'm curious if we'd be better off assuming that a reset is a wrap instead. I'm sure there's literature on the topic. It can wait for later though.

Reply:
++